File: lib/data/providers/community_wrapper_provider.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import '../../domain/entities/mosque.dart';
import '../../domain/entities/prayer_times.dart';
import '../../domain/entities/geo_location.dart';
import 'prayer_data_provider.dart';

/// Provider B: Community Wrapper API
/// User-supplied REST wrapper endpoint
class CommunityWrapperProvider implements PrayerDataProvider {
  late Dio _dio;
  String? _baseUrl;
  String? _apiKey;
  
  @override
  String get id => 'community_wrapper';
  
  @override
  String get name => 'Community API Wrapper';
  
  @override
  String get description => 'Third-party REST wrapper for Mawaqit data';
  
  @override
  List<ConfigField> get configSchema => [
    const ConfigField(
      key: 'baseUrl',
      label: 'Wrapper Base URL',
      type: ConfigFieldType.url,
      required: true,
      description: 'e.g., https://mawaqit-wrapper.example.com',
    ),
    const ConfigField(
      key: 'apiKey',
      label: 'API Key (optional)',
      type: ConfigFieldType.password,
      required: false,
      description: 'If your wrapper requires authentication',
    ),
    const ConfigField(
      key: 'timeout',
      label: 'Request Timeout (seconds)',
      type: ConfigFieldType.number,
      defaultValue: '30',
    ),
  ];

  @override
  Future<void> initialize(Map<String, dynamic> config) async {
    _baseUrl = config['baseUrl'] as String?;
    _apiKey = config['apiKey'] as String?;
    final timeout = int.tryParse(config['timeout']?.toString() ?? '30') ?? 30;
    
    if (_baseUrl == null) {
      throw const ProviderException(
        message: 'Base URL is required',
        providerId: 'community_wrapper',
      );
    }

    // Ensure baseUrl doesn't end with /
    final baseUrl = _baseUrl!.endsWith('/') ? _baseUrl!.substring(0, _baseUrl!.length - 1) : _baseUrl;

    final headers = <String, String>{
      'Accept': 'application/json',
      'User-Agent': 'IqamahApp/1.0',
    };

    if (_apiKey != null && _apiKey!.isNotEmpty) {
      headers['X-API-Key'] = _apiKey!;
    }

    _dio = Dio(BaseOptions(
      baseUrl: baseUrl!,
      connectTimeout: Duration(seconds: timeout),
      receiveTimeout: Duration(seconds: timeout),
      headers: headers,
    ));
  }

  @override
  Future<List<Mosque>> searchMosques(String query, {GeoLocation? location}) async {
    try {
      final response = await _dio.get('/mosques', queryParameters: {
        'search': query,
        if (location != null) ...{
          'lat': location.latitude,
          'lng': location.longitude,
        },
      });

      if (response.statusCode == 200) {
        final List<dynamic> data;
        if (response.data is List) {
          data = response.data as List;
        } else if (response.data is Map) {
          data = (response.data as Map)['data'] ?? (response.data as Map)['mosques'] ?? (response.data as Map)['results'] ?? [];
        } else {
          data = [];
        }
        return data.map((json) => _mapMosque(json)).toList();
      }
      
      throw ProviderException(
        message: 'Failed to search mosques: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<List<Mosque>> getNearbyMosques(GeoLocation location, {double radiusKm = 10}) async {
    try {
      final response = await _dio.get('/mosques/nearby', queryParameters: {
        'lat': location.latitude,
        'lng': location.longitude,
        'radius': radiusKm,
      });

      if (response.statusCode == 200) {
        final List<dynamic> data;
        if (response.data is List) {
          data = response.data as List;
        } else {
          data = (response.data as Map)['data'] ?? (response.data as Map)['mosques'] ?? [];
        }
        return data.map((json) => _mapMosque(json)).toList();
      }
      
      throw ProviderException(
        message: 'Failed to get nearby mosques: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<PrayerTimes> getPrayerTimes(String mosqueId, {DateTime? date}) async {
    try {
      final targetDate = date ?? DateTime.now();
      final response = await _dio.get('/mosques/$mosqueId/times', queryParameters: {
        'date': '${targetDate.year}-${targetDate.month.toString().padLeft(2, '0')}-${targetDate.day.toString().padLeft(2, '0')}',
      });

      if (response.statusCode == 200) {
        return _mapPrayerTimes(response.data, mosqueId, targetDate);
      }
      
      throw ProviderException(
        message: 'Failed to get prayer times: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<Mosque> getMosqueDetails(String mosqueId) async {
    try {
      final response = await _dio.get('/mosques/$mosqueId');

      if (response.statusCode == 200) {
        return _mapMosque(response.data);
      }
      
      throw ProviderException(
        message: 'Failed to get mosque details: ${response.statusCode}',
        statusCode: response.statusCode,
        providerId: id,
      );
    } on DioException catch (e) {
      throw ProviderException(
        message: 'Network error: ${e.message}',
        statusCode: e.response?.statusCode,
        providerId: id,
      );
    }
  }

  @override
  Future<bool> testConnection() async {
    try {
      final response = await _dio.get('/health');
      return response.statusCode == 200;
    } catch (e) {
      // Try alternative endpoints
      try {
        final response = await _dio.get('/');
        return response.statusCode == 200;
      } catch (_) {
        return false;
      }
    }
  }

  Mosque _mapMosque(Map<String, dynamic> json) {
    // Handle various field naming conventions
    return Mosque(
      id: json['id']?.toString() ?? json['mosque_id']?.toString() ?? '',
      name: json['name'] ?? json['mosque_name'] ?? json['title'] ?? 'Unknown Mosque',
      address: json['address'] ?? json['location'] ?? json['street'] ?? json['address_line'],
      city: json['city'] ?? json['ville'] ?? json['town'],
      country: json['country'] ?? json['pays'] ?? json['nation'],
      latitude: _parseDouble(json['latitude'] ?? json['lat'] ?? json['latitude_deg']),
      longitude: _parseDouble(json['longitude'] ?? json['lng'] ?? json['lon'] ?? json['longitude_deg']),
    );
  }

  double? _parseDouble(dynamic value) {
    if (value == null) return null;
    if (value is double) return value;
    if (value is int) return value.toDouble();
    if (value is String) return double.tryParse(value);
    return null;
  }

  PrayerTimes _mapPrayerTimes(dynamic data, String mosqueId, DateTime date) {
    Map<String, dynamic> times;
    
    if (data is Map) {
      times = (data['times'] ?? data['prayer_times'] ?? data['schedule'] ?? data) as Map<String, dynamic>;
    } else {
      times = {};
    }

    return PrayerTimes(
      date: date,
      mosqueId: mosqueId,
      cachedAt: DateTime.now(),
      fajr: _parsePrayer('Fajr', times['fajr'] ?? times['Fajr'] ?? times['sobh'] ?? times['fajr_adhan'], date),
      dhuhr: _parsePrayer('Dhuhr', times['dhuhr'] ?? times['Dhuhr'] ?? times['duhr'] ?? times['dhuhr_adhan'] ?? times['zuhr'], date),
      asr: _parsePrayer('Asr', times['asr'] ?? times['Asr'] ?? times['asr_adhan'], date),
      maghrib: _parsePrayer('Maghrib', times['maghrib'] ?? times['Maghrib'] ?? times['maghrib_adhan'] ?? times['magreb'], date),
      isha: _parsePrayer('Isha', times['isha'] ?? times['Isha'] ?? times['isha_adhan'], date),
      jumuah: times['jumuah'] != null || times['Jumuah'] != null || times['juma'] != null
        ? _parsePrayer('Jumuah', times['jumuah'] ?? times['Jumuah'] ?? times['juma'], date)
        : null,
    );
  }

  Prayer _parsePrayer(String name, dynamic data, DateTime baseDate) {
    if (data == null) {
      return Prayer(name: name, adhan: baseDate, iqama: null);
    }

    if (data is Map) {
      return Prayer(
        name: name,
        adhan: _parseTime(data['adhan'] ?? data['athan'] ?? data['azan'] ?? data['time'] ?? data['start'] ?? data, baseDate),
        iqama: data['iqama'] != null || data['iqamah'] != null || data['jamaah'] != null
          ? _parseTime(data['iqama'] ?? data['iqamah'] ?? data['jamaah'], baseDate)
          : null,
      );
    }

    return Prayer(
      name: name,
      adhan: _parseTime(data, baseDate),
      iqama: null,
    );
  }

  DateTime _parseTime(dynamic timeData, DateTime baseDate) {
    if (timeData is DateTime) return timeData;
    
    final timeStr = timeData.toString().trim();
    
    // Handle HH:MM format
    final timeParts = timeStr.split(':');
    if (timeParts.length >= 2) {
      var hour = int.tryParse(timeParts[0]) ?? 0;
      final minute = int.tryParse(timeParts[1]) ?? 0;
      
      // Handle 12-hour format with AM/PM
      if (timeStr.toLowerCase().contains('pm') && hour != 12) {
        hour += 12;
      } else if (timeStr.toLowerCase().contains('am') && hour == 12) {
        hour = 0;
      }
      
      return DateTime(baseDate.year, baseDate.month, baseDate.day, hour, minute);
    }
    
    return baseDate;
  }

  @override
  void dispose() {
    _dio.close();
  }
}

File: src-tauri/src/providers/community_wrapper_provider.rs
use async_trait::async_trait;
use chrono::{NaiveDate, Utc};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::models::*;
use crate::providers::{PrayerDataProvider, ProviderError, ProviderResult};

/// Community Wrapper API Provider (Provider B)
/// REST API wrapper that provides a simplified interface
pub struct CommunityWrapperProvider {
    client: Client,
    base_url: Option<String>,
    api_key: Option<String>,
}

impl CommunityWrapperProvider {
    pub fn new() -> Self {
        Self {
            client: Client::new(),
            base_url: None,
            api_key: None,
        }
    }

    fn ensure_initialized(&self) -> ProviderResult<(String, Option<String>)> {
        match &self.base_url {
            Some(url) => Ok((url.clone(), self.api_key.clone())),
            None => Err(ProviderError::InvalidConfig(
                "Provider not initialized".to_string(),
            )),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct MosqueResponse {
    id: String,
    name: String,
    address: Option<String>,
    city: Option<String>,
    country: Option<String>,
    latitude: Option<f64>,
    longitude: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct PrayerTimesResponse {
    date: String,
    fajr: PrayerResponse,
    dhuhr: PrayerResponse,
    asr: PrayerResponse,
    maghrib: PrayerResponse,
    isha: PrayerResponse,
    jumuah: Option<PrayerResponse>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct PrayerResponse {
    adhan: String,
    iqama: Option<String>,
    rakah_count: Option<i32>,
}

#[async_trait]
impl PrayerDataProvider for CommunityWrapperProvider {
    fn id(&self) -> &str {
        PROVIDER_COMMUNITY_WRAPPER
    }

    fn name(&self) -> &str {
        "Community Wrapper API"
    }

    fn description(&self) -> &str {
        "Community-provided REST API wrapper for prayer times data"
    }

    fn config_schema(&self) -> Vec<ConfigField> {
        vec![
            ConfigField::new("base_url", "API Base URL", ConfigFieldType::Url)
                .required()
                .description("The base URL of the community API"),
            ConfigField::new("api_key", "API Key", ConfigFieldType::Password)
                .description("Optional API key for authentication"),
        ]
    }

    async fn initialize(&mut self, config: Value) -> ProviderResult<()> {
        self.base_url = config["base_url"]
            .as_str()
            .map(|s| s.trim_end_matches('/').to_string());
        self.api_key = config["api_key"].as_str().map(|s| s.to_string());
        Ok(())
    }

    async fn search_mosques(&self, query: &str, _location: Option<&GeoLocation>) -> ProviderResult<Vec<Mosque>> {
        let (base_url, api_key) = self.ensure_initialized()?;

        let mut request = self.client.get(format!("{}/mosques/search", base_url));
        request = request.query(&[("q", query)]);

        if let Some(key) = api_key {
            request = request.header("X-API-Key", key);
        }

        let response = request.send().await.map_err(|e| {
            ProviderError::Network(format!("Failed to search mosques: {}", e))
        })?;

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Failed to search mosques".to_string(),
            });
        }

        let mosques: Vec<MosqueResponse> = response.json().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to parse response: {}", e))
        })?;

        Ok(mosques
            .into_iter()
            .map(|m| Mosque {
                id: m.id,
                name: m.name,
                address: m.address,
                city: m.city,
                country: m.country,
                latitude: m.latitude,
                longitude: m.longitude,
                is_favorite: false,
                last_accessed: None,
            })
            .collect())
    }

    async fn get_nearby_mosques(&self, location: &GeoLocation, radius_km: f64) -> ProviderResult<Vec<Mosque>> {
        let (base_url, api_key) = self.ensure_initialized()?;

        let mut request = self.client.get(format!("{}/mosques/nearby", base_url));
        request = request.query(&[
            ("lat", location.latitude.to_string()),
            ("lng", location.longitude.to_string()),
            ("radius", radius_km.to_string()),
        ]);

        if let Some(key) = api_key {
            request = request.header("X-API-Key", key);
        }

        let response = request.send().await.map_err(|e| {
            ProviderError::Network(format!("Failed to get nearby mosques: {}", e))
        })?;

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Failed to get nearby mosques".to_string(),
            });
        }

        let mosques: Vec<MosqueResponse> = response.json().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to parse response: {}", e))
        })?;

        Ok(mosques
            .into_iter()
            .map(|m| Mosque {
                id: m.id,
                name: m.name,
                address: m.address,
                city: m.city,
                country: m.country,
                latitude: m.latitude,
                longitude: m.longitude,
                is_favorite: false,
                last_accessed: None,
            })
            .collect())
    }

    async fn get_prayer_times(&self, mosque_id: &str, date: Option<NaiveDate>) -> ProviderResult<PrayerTimes> {
        let (base_url, api_key) = self.ensure_initialized()?;

        let mut request = self
            .client
            .get(format!("{}/mosques/{}/prayer-times", base_url, mosque_id));

        if let Some(d) = date {
            request = request.query(&[("date", d.format("%Y-%m-%d").to_string())]);
        }

        if let Some(key) = api_key {
            request = request.header("X-API-Key", key);
        }

        let response = request.send().await.map_err(|e| {
            ProviderError::Network(format!("Failed to get prayer times: {}", e))
        })?;

        if response.status().as_u16() == 404 {
            return Err(ProviderError::NotFound(format!(
                "Mosque {} not found",
                mosque_id
            )));
        }

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Failed to get prayer times".to_string(),
            });
        }

        let times: PrayerTimesResponse = response.json().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to parse response: {}", e))
        })?;

        let date = NaiveDate::parse_from_str(&times.date, "%Y-%m-%d")
            .map_err(|e| ProviderError::Parse(format!("Invalid date: {}", e)))?
            .and_hms_opt(0, 0, 0)
            .unwrap()
            .and_local_timezone(Utc)
            .unwrap();

        fn parse_prayer(name: &str, p: PrayerResponse, date: chrono::DateTime<Utc>) -> Prayer {
            let time_part = |t: &str| {
                let parts: Vec<&str> = t.split(':').collect();
                let hour: u32 = parts[0].parse().unwrap_or(0);
                let minute: u32 = parts.get(1).unwrap_or(&"0").parse().unwrap_or(0);
                date.date_naive().and_hms_opt(hour, minute, 0).unwrap().and_local_timezone(Utc).unwrap()
            };

            Prayer {
                name: name.to_string(),
                adhan: time_part(&p.adhan),
                iqama: p.iqama.as_ref().map(|t| time_part(t)),
                custom_rakah_count: p.rakah_count,
            }
        }

        Ok(PrayerTimes {
            date,
            fajr: parse_prayer("Fajr", times.fajr, date),
            dhuhr: parse_prayer("Dhuhr", times.dhuhr, date),
            asr: parse_prayer("Asr", times.asr, date),
            maghrib: parse_prayer("Maghrib", times.maghrib, date),
            isha: parse_prayer("Isha", times.isha, date),
            jumuah: times.jumuah.map(|j| parse_prayer("Jumuah", j, date)),
            mosque_id: Some(mosque_id.to_string()),
            mosque_name: None,
            cached_at: Some(Utc::now()),
        })
    }

    async fn test_connection(&self) -> ProviderResult<ProviderTestResult> {
        let start = std::time::Instant::now();

        let (base_url, api_key) = self.ensure_initialized()?;

        let mut request = self.client.get(format!("{}/health", base_url));

        if let Some(key) = api_key {
            request = request.header("X-API-Key", key);
        }

        match request.send().await {
            Ok(response) => {
                let latency = start.elapsed().as_millis() as u64;
                if response.status().is_success() {
                    Ok(ProviderTestResult {
                        success: true,
                        message: "Connection successful".to_string(),
                        latency_ms: Some(latency),
                    })
                } else {
                    Ok(ProviderTestResult {
                        success: false,
                        message: format!("Server returned: {}", response.status()),
                        latency_ms: Some(latency),
                    })
                }
            }
            Err(e) => Ok(ProviderTestResult {
                success: false,
                message: format!("Connection failed: {}", e),
                latency_ms: None,
            }),
        }
    }

    async fn get_mosque_details(&self, mosque_id: &str) -> ProviderResult<Mosque> {
        let (base_url, api_key) = self.ensure_initialized()?;

        let mut request = self
            .client
            .get(format!("{}/mosques/{}", base_url, mosque_id));

        if let Some(key) = api_key {
            request = request.header("X-API-Key", key);
        }

        let response = request.send().await.map_err(|e| {
            ProviderError::Network(format!("Failed to get mosque details: {}", e))
        })?;

        if response.status().as_u16() == 404 {
            return Err(ProviderError::NotFound(format!(
                "Mosque {} not found",
                mosque_id
            )));
        }

        if !response.status().is_success() {
            return Err(ProviderError::Server {
                status_code: response.status().as_u16(),
                message: "Failed to get mosque details".to_string(),
            });
        }

        let m: MosqueResponse = response.json().await.map_err(|e| {
            ProviderError::Parse(format!("Failed to parse response: {}", e))
        })?;

        Ok(Mosque {
            id: m.id,
            name: m.name,
            address: m.address,
            city: m.city,
            country: m.country,
            latitude: m.latitude,
            longitude: m.longitude,
            is_favorite: false,
            last_accessed: None,
        })
    }
}
